### PR TITLE
Use uv.

### LINKED ISSUE #9868: Use uv
Replace the following tools with uv:
- [x] pip
- [x] pipx: `uv tool run`
- [x] pip-compile

### PR DESCRIPTION
**Pull request by fniessink**
Note that `uv` names the virtual env folder `.venv` while we have been using `venv` as default. Developers may have to recreate the virtual environments on their laptop.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -24,8 +24,8 @@ jobs:
           esac
           cd $component
           mkdir -p build
-          python -m venv venv
-          . venv/bin/activate
```
Comment (wkoot) [inline .circleci/config.yml:28]: Why do we not want to work within a virtual env anymore, but only call the binary within the venv?

File: .circleci/config.yml [hunk 2]
```diff
@@ -91,7 +90,7 @@ jobs:
           export COMPOSE_FILE=docker/docker-compose.yml:docker/docker-compose.ci.yml
           docker compose build && docker compose up -d
           docker ps
-          docker run -it -w `pwd` -v `pwd`:`pwd` --network=container:docker-www-1 python:3.12.5-bookworm tests/application_tests/ci/test.sh
+          docker run -it -w `pwd` -v `pwd`:`pwd` --network=container:docker-www-1 ghcr.io/astral-sh/uv:python3.12-bookworm tests/application_tests/ci/test.sh
```
Comment (wkoot) [inline .circleci/config.yml:93]: Is this a, specifically for uv, altered python container, or just an alternative registry backed by github?

File: .circleci/config.yml [hunk 3]
```diff
@@ -110,6 +109,7 @@ jobs:
     steps:
       - checkout
       - run: |
+          pip install uv
```
Comment (wkoot) [inline .circleci/config.yml:112]: Is there no `uv venv` needed?

File: .github/workflows/api_server.yml [hunk 4]
```diff
@@ -5,24 +5,28 @@ on: [push]
...
-      run: |
-        cd components/api_server
-        ci/quality.sh
+      - uses: actions/checkout@v4.2.0
+      - name: Install uv
+        uses: astral-sh/setup-uv@v3
+        with:
+          version: "0.4.18"
+      - name: "Set up Python"
+        uses: actions/setup-python@v5
```
Comment (wkoot) [inline .github/workflows/api_server.yml]: Why do we need setup python when we already install uv, or does this action use the uv environment?

File: .github/workflows/api_server.yml [hunk 5]
```diff
@@ -5,24 +5,28 @@ on: [push]
...
+        with:
+          python-version-file: "components/api_server/pyproject.toml"
+      - name: Install dependencies
```
Comment (wkoot) [inline .github/workflows/api_server.yml]: Why is this needed when the previous step already used the `pyproject.toml` to set up the environment?

File: .github/workflows/application-tests.yml [hunk 6]
```diff
@@ -17,7 +17,7 @@ jobs:
         run: |
           mkdir -p build
           docker compose --file docker/docker-compose.yml --file docker/docker-compose.ci.yml --project-name quality-time up --build --detach --wait
-          docker run -t -w `pwd` -v `pwd`:`pwd` --network=container:quality-time-www-1 python:3.12.5-bookworm tests/application_tests/ci/test.sh
+          docker run -t -w `pwd` -v `pwd`:`pwd` --network=container:quality-time-www-1 ghcr.io/astral-sh/uv:python3.12-bookworm tests/application_tests/ci/test.sh
```
Comment (wkoot) [inline .github/workflows/application-tests.yml:20]: Do we still need debian, or can we go to alpine?

File: .github/workflows/notifier.yml [hunk 7]
```diff
@@ -5,24 +5,28 @@ on: [push]
...
-        ci/quality.sh
+      - uses: actions/checkout@v4.2.0
+      - name: Install uv
+        uses: astral-sh/setup-uv@v3
+        with:
+          version: "0.4.18"
+      - name: "Set up Python"
+        uses: actions/setup-python@v5
+        with:
+          python-version-file: "components/collector/pyproject.toml"
```
Comment (wkoot) [inline .github/workflows/notifier.yml]:
```suggestion
          python-version-file: "components/notifier/pyproject.toml"
```

File: .github/workflows/release-quality.yml [hunk 8]
```diff
@@ -5,20 +5,24 @@ on: [push]
...
-        ci/quality.sh
+      - uses: actions/checkout@v4.2.0
+      - name: Install uv
+        uses: astral-sh/setup-uv@v3
+        with:
+          version: "0.4.18"
+      - name: "Set up Python"
+        uses: actions/setup-python@v5
+        with:
+          python-version-file: "components/collector/pyproject.toml"
```
Comment (wkoot) [inline .github/workflows/release-quality.yml]:
```suggestion
          python-version-file: "release/pyproject.toml"
```

File: .github/workflows/shared_code.yml [hunk 9]
```diff
@@ -5,24 +5,28 @@ on: [push]
...
-        ci/quality.sh
+      - uses: actions/checkout@v4.2.0
+      - name: Install uv
+        uses: astral-sh/setup-uv@v3
+        with:
+          version: "0.4.18"
+      - name: "Set up Python"
+        uses: actions/setup-python@v5
+        with:
+          python-version-file: "components/collector/pyproject.toml"
```
Comment (wkoot) [inline .github/workflows/shared_code.yml]:
```suggestion
          python-version-file: "components/shared_code/pyproject.toml"
```

File: .github/workflows/sonarcloud.yml [hunk 10]
```diff
@@ -18,11 +18,16 @@ jobs:
         if: env.SONAR_TOKEN != null
         with:
           fetch-depth: 0 # Disabling shallow clone is recommended for improving relevancy of reporting
-      - name: Set up Python
+      - name: Install uv
+        uses: astral-sh/setup-uv@v3
+        if: env.SONAR_TOKEN != null
```
Comment (wkoot) [inline .github/workflows/sonarcloud.yml:22]: Can't we add this conditional on the entire workflow, or step 1 which aborts the workflow?

File: .github/workflows/sonarcloud.yml [hunk 11]
```diff
@@ -18,11 +18,16 @@ jobs:
...
+        with:
+          version: "0.4.18"
+      - name: "Set up Python"
         uses: actions/setup-python@v5
         if: env.SONAR_TOKEN != null
         with:
-          python-version: "3.12"
+          python-version-file: "components/collector/pyproject.toml"
```
Comment (wkoot) [inline .github/workflows/sonarcloud.yml]: Oh, is this really only pulling the python version from the toml!

File: ci/uvx-base.sh [hunk 12]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
+
+source base.sh
+
+run_uvx() {
+    # Look up the version of the command using the spec function and run the command using uvx.
+    command_spec=$(spec $1 @)
```
Comment (wkoot) [inline ci/uvx-base.sh:7]: I don't get it, where does the `@` come from? This produces something like `package@v1.2.3` - is that the syntax for uvx?
Also note that the pypi page for uvx links to a `uvx2` repo, which is apparently renamed to `uvenv` to avoid naming collision with a new uv command

File: ci/spec.py [hunk 13]
```diff
@@ -14,8 +14,8 @@ def spec(package: str, pyproject_toml_path: Path) -> str:
         pyproject_toml = tomllib.load(pyproject_toml_file)
     tools = pyproject_toml["project"]["optional-dependencies"]["tools"]
     package_specs = [spec for spec in tools if spec.split("==")[0] == package]
-    return package_specs[0] if package_specs else ""
+    return sep.join(package_specs[0].split("==")) if package_specs else ""
```
Comment (wkoot) [inline ci/spec.py:17]: This method is kinda dirty, can't we comprehend the toml file directly? The `tomllib` is a std lib since python 3.11

File: tests/feature_tests/ci/test.sh [hunk 14]
```diff
@@ -11,34 +11,31 @@ export PROXY_PORT=8080
...
 cd tests/feature_tests
-python3 -m venv venv
-. venv/bin/activate
+uv venv
 ci/pip-install.sh
 cd ../..
 sleep 5  # Give components time to start up
-coverage erase
-coverage run -m behave --format pretty "${1:-tests/feature_tests/src/features}"
+tests/feature_tests/.venv/bin/coverage erase
```
Comment (wkoot) [inline tests/feature_tests/ci/test.sh:28]: Why do we `cd ../..` above, when everything is then calling the previous path? If it's only for the combine step, perhaps that one can be prefixed?

File: ci/uvx-base.sh [hunk 15]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
+
+source base.sh
+
+run_uvx() {
```
Comment (wkoot) [inline ci/uvx-base.sh:5]: Note that the `uvx` project has been renamed to `uvenv` due to naming collision

File: ci/quality-base.sh [hunk 16]
```diff
@@ -1,55 +1,54 @@
...
-    run_pipx fixit lint $(python_files_and_folders)
+    run_uvx fixit lint $(python_files_and_folders)
 }
 
 run_mypy() {
     PYTHON_FILES_AND_FOLDERS=$(python_files_and_folders)
     # Run mypy with or without pydantic plugin depending on whether pydantic is listed as dependency in the tools
     # section of the optional dependencies in the pyproject.toml file.
-    pydantic_spec=$(spec pydantic)
+    pydantic_spec=$(spec pydantic ==)
```
Comment (wkoot) [inline ci/quality-base.sh:23]: This seems to be the only place where spec is called with a `==` separator, why the distinction?

File: ci/quality-base.sh [hunk 17]
```diff
@@ -1,55 +1,54 @@
...
     if [[ "$pydantic_spec" == "" ]]; then
-        run_pipx mypy --python-executable=$(which python) $PYTHON_FILES_AND_FOLDERS
+        run_uvx mypy --python-executable=.venv/bin/python $PYTHON_FILES_AND_FOLDERS
     else
-        # To use the pydantic plugin, we need to first install mypy and then inject pydantic
-        run pipx install --force $(spec mypy) # --force works around this bug: https://github.com/pypa/pipx/issues/795
-        run pipx inject mypy $pydantic_spec
-        run $PIPX_BIN_DIR/mypy --python-executable=$(which python) $PYTHON_FILES_AND_FOLDERS
+        run uvx --with $pydantic_spec $(spec mypy @) --python-executable=.venv/bin/python $PYTHON_FILES_AND_FOLDERS
```
Comment (wkoot) [inline ci/quality-base.sh:27]: I don't get it, doesn't this produce something like `--with pydantic==1.2.3 mypy@4.5.6` ?

File: .circleci/config.yml (hunk 1 above)
Comment (fniessink) [reply .circleci/config.yml:28]: `uv venv` creates the virtual env.

File: .circleci/config.yml (hunk 2 above)
Comment (fniessink) [reply .circleci/config.yml:93]: This is a Python 3.12 container with uv preinstalled. We could use a vanilla Python container, but then the `test.sh` script would have to install uv.

File: .circleci/config.yml (hunk 3 above)
Comment (fniessink) [reply .circleci/config.yml:112]: No, the `test.sh` script creates the virtual env.

File: .github/workflows/api_server.yml (hunk 4 above)
Comment (fniessink) [reply .github/workflows/api_server.yml]: Good question. I'll test without it and see if uv correctly downloads Python.

Comment (fniessink) [reply .github/workflows/api_server.yml]: That seems to work. Nice.

File: .github/workflows/api_server.yml (hunk 5 above)
Comment (fniessink) [reply .github/workflows/api_server.yml]: Indeed not needed.

File: .github/workflows/application-tests.yml (hunk 6 above)
Comment (fniessink) [reply .github/workflows/application-tests.yml:20]: I'll test.

Comment (fniessink) [reply .github/workflows/application-tests.yml:20]: Alas: `exec tests/application_tests/ci/test.sh: no such file or directory`.

File: .github/workflows/sonarcloud.yml (hunk 10 above)
Comment (fniessink) [reply .github/workflows/sonarcloud.yml:22]: Testing...

File: ci/uvx-base.sh (hunk 12 above)
Comment (fniessink) [reply ci/uvx-base.sh:7]: uvx is a bit inconsistent with the package spec, it uses both @ and ==: `uvx --with package1==a.b.c package2@x.y.z`. Hence the need to pass the separator to the `spec` function.

Note that `uvx` is not a separate package but an alias for `uv tool run`.

File: ci/uvx-base.sh (hunk 15 above)
Comment (fniessink) [reply ci/uvx-base.sh:5]: We don't use the uvx project.

File: ci/spec.py (hunk 13 above)
Comment (fniessink) [reply ci/spec.py:17]: We use tomllib to read the toml file, but the package specs are strings of the form `"package==x.y.z"`. tomllib doesn't help with that.

File: tests/feature_tests/ci/test.sh (hunk 14 above)
Comment (fniessink) [reply tests/feature_tests/ci/test.sh:28]: I don't remember exactly but it's probably to make sure the coverage files don't end up in `tests/feature_tests`.

File: ci/quality-base.sh (hunk 16 above)
Comment (fniessink) [reply ci/quality-base.sh:23]: See above.

File: ci/quality-base.sh (hunk 17 above)
Comment (fniessink) [reply ci/quality-base.sh:27]: Yup!

File: .github/workflows/sonarcloud.yml (hunk 10 above)
Comment (fniessink) [reply .github/workflows/sonarcloud.yml:22]: Nope. Workflow level environment variables aren't available in the job: https://github.com/actions/runner/issues/1661.

File: .circleci/config.yml (hunk 1 above)
Comment (wkoot) [reply .circleci/config.yml:28]: Yes but `source .venv/bin/activate` would still source that newly created venv

File: .github/workflows/application-tests.yml (hunk 6 above)
Comment (wkoot) [reply .github/workflows/application-tests.yml:20]: Hmm, where does it go wrong? Does it lack a bash interpreter perhaps? If so, `RUN apk add --no-cache bash` should help

File: ci/uvx-base.sh (hunk 12 above)
Comment (wkoot) [reply ci/uvx-base.sh:7]: ouch, no wonder I couldn't understand the docs!

File: .circleci/config.yml (hunk 1 above)
Comment (fniessink) [reply .circleci/config.yml:28]: Yes, but `ci/pip-install.sh` runs `uv pip install` which uses the venv without activation. See https://docs.astral.sh/uv/pip/environments/#discovery-of-python-environments.

`ci/unittest.sh` runs `.venv/bin/coverage` (I'll have to check whether I can get `uvx coverage` working) and `ci/quality.sh` run everything via uvx so they don't need an activated virtual env either.

I think we're very close to never having to activate a virtual envs anymore. The only thing I haven't found out is how to run the application (e.g. `python src/quality_time_server.py`) without activating the venv. `uv run src/quality_time_server.py` insists on creating a `uv.lock` file which we don't need as we have compiled requirements files.

Comment (fniessink) [reply .circleci/config.yml:28]: > I'll have to check whether I can get `uvx coverage` working

Alas, that isn't possible. `mypy` has a `--python-executable` flag for this purpose, but `coverage` does not.
